keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,554 @@
+- Immutability
+- Unions
+
+These improvements are largely complementary to expression support and either
+are (or should be) addressed by in separate KEPs.
+
+For use cases cannot be covered by build-in validation support:
+
+- Admission Webhooks: have validating admission webhook for further validation
+- Custom validators: write custom checks in several languages such as Rego 
+
+### Descriptive, self contained CRDs
+
+This KEP will make CRDs more self-contained. Instead of having
+validation rules coded into webhooks that must be
+registered and upgraded independent of a CRD, the rules will be contained within
+the CRD object definition, making them easier to author and introspect by
+cluster administrators and users, and eliminating version skew issues that can
+happen between a CRD and webhook since they can be registered and
+upgraded/rolled-back independently.
+
+### Webhooks: Development Complexity
+
+Introducing a production grade webhook is a substantial development task.
+Beyond authoring the actual core logic that a webhook must perform, the webhook
+must be instrumented for monitoring and alerting and integrated with the
+packaging/releases processes for the environments it will be run it.
+
+The developer must also carefully consider the upgrade and rollback ordering
+between the webhook and CRD.
+
+### Webhooks: Operational Complexity
+
+Admission webhooks are part of the critical serving path of the kube-apiserver.
+Admission webhooks add latency to requests, and large numbers of webhooks
+can cause, or contribute to, request timeouts being exceeded.
+
+Webhooks must either be configured as `FailPolicy.Fail` or `FailPolicy.Ignore`. If
+`FailPolicy.Ignore` is used, there is potential for requests skip the webhook and
+be admitted. If `FailPolicy.Fail` is used, a webhook outage can result in a
+localized or widespread Kubernetes control plane outage depending on which
+objects the webhook is configured to intercept.
+
+### Goals
+
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+- Support for validation formats, immutability or unions. These are all valuable improvements
+  but can be addressed orthogonally in separate KEPs.
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. For example, if a validation check requires making a network
+  request to some other system, it will still need to be implemented in a webhook.
+- Change how Kubernetes built-in types are validated, defaulted and converted.
+
+## Proposal
+
+An inline expression language like [Common Expression Language (CEL)](https://github.com/google/cel-go) would be an
+excellent supplement to the current validation mechanism because it is sufficiently expressive to
+satisfy a large set of remaining uses cases that none of the above can solve.
+For example, cross-field validation use cases can only be solved using
+expressions or webhooks.
```

Comment:
Thanks for pointing out. :)

Reply:
I've added a couple notes to the KEP to make the CEL limitations more concrete. CRD structural schemas are a great example.

Reply:
Oh, I should mention-- Since CEL allows us to provide our own builtin library of functions. If we ever did want to make it possible to validate structural schemas, we could provide a `isStructuralSchema` build in function (the implementation of the function would be in go).  I don't think we need that now, just wanted to point out we do have the build in function library as an "escape hatch" if we have some specific thing we need to do that CEL cannot support.

Reply:
> we do have the build in function library as an "escape hatch" if we have some specific thing we need to do that CEL cannot support.

that's true, but it's important to keep in mind that every time we add a new built-in macro/function, we limit version compatibility for any CRD making use of it

Reply:
Very true. I've added a section to the graduation criteria (for beta) about this macro/function library.